pkg/kvstore/etcd.go

```diff
@@ -656,30 +656,18 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
 
 		go ec.statusChecker()
 
-		watcher := ec.ListAndWatch(ctx, HeartbeatPath, 128)
-
-		for {
-			select {
-			case _, ok := <-watcher.Events:
-				if !ok {
-					ec.logger.Debug("Stopping heartbeat watcher")
-					watcher.Stop()
-					return
-				}
-
-				// It is tempting to compare against the
-				// heartbeat value stored in the key. However,
-				// this would require the time on all nodes to
-				// be synchronized. Instead, assume current
-				// time and print the heartbeat value in debug
-				// messages for troubleshooting
-				ec.RWMutex.Lock()
-				ec.lastHeartbeat = time.Now()
-				ec.RWMutex.Unlock()
-				ec.logger.Debug("Received update notification of heartbeat")
-			case <-ctx.Done():
-				return
-			}
+		watcher := ec.ListAndWatch(ctx, HeartbeatPath, 0)
+		for range watcher.Events {
```

Comment:
I think it depends, as even an unrelated event is still a liveness signal (at least of the watch stream).

Considering the current implementation, considering the list done event is actually useful to set the last heartbeat variable to a non-zero value, which then actually enables the check. Otherwise, it would be skipped due to:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/pkg/kvstore/etcd.go#L1034-L1037

I instead agree that a delete event could be instead safely ignored.

Reply:
:thinking: :thinking: :thinking: 
Sorry, I don't understand it. 

ListDone is an artificially created event [here](https://github.com/cilium/cilium/blob/f30d83a2476adaae74cdcf27a395e3c48d258399/pkg/kvstore/etcd.go#L872)

I would always assume that heartbeat events correspond to something that happened in relation to the outside world.

> Considering the current implementation, considering the list done event is actually useful to set the last heartbeat variable to a non-zero value, which then actually enables the check. Otherwise, it would be skipped due to:

these all checks are very suspicious. If we never create a watch and keep relisting the same key, it would still never fail, no?

imho we should be checking if heartbeat value changed since last time we got it and only then update `lastHeartbeat`.

Reply:
> ListDone is an artificially created event [here](https://github.com/cilium/cilium/blob/f30d83a2476adaae74cdcf27a395e3c48d258399/pkg/kvstore/etcd.go#L872)

Yes, my point is that it is a signal that we actually successfully listed the heartbeat key (which may, or may not exist at this time). Setting the `lastHeartbeat` variable in that case ensures that if we don't receive any event within the expected timeout, we'll consider that as an heartbeat failure, which looks the correct behavior to me. Especially considering that with the new changes the status checker is only started once receiving the list done event (which again looks correct to me, as we want to run the status checker only once exited the initial connection phase, as guarded by a separate timeout).

> these all checks are very suspicious. If we never create a watch and keep relisting the same key, it would still never fail, no?

Sort of. That's what happened when I accidentally dropped the permissions for the heartbeat prefix, preventing it from starting. This specific case is now prevented as the list operation needs to succeed before considering the client ready. There's definitely the possibility that the watch keeps subsequently failing, triggering list operation agains, and the corresponding upsert event, which would incorrectly count as an heartbeat. Although arguably that looks extremely rare to me, considering that it would require the watcher to continuously fail due to a compacted error.

That said, I totally agree with you that the status checker logic could be improved to address these corner cases and make the overall logic more straightforward to reason about. Although I'm always a bit scared of touching it due to the possible unforeseen consequences on Cilium running in kvstore mode, as the lifecycle there is completely different.

Reply:
> Yes, my point is that it is a signal that we actually successfully listed the heartbeat key (which may, or may not exist at this time). Setting the lastHeartbeat variable in that case ensures that if we don't receive any event within the expected timeout, we'll consider that as an heartbeat failure, which looks the correct behavior to me.

I agree this is correct behavior, but it strongly suggests an incorrect division of responsibilities between heartbeat watcher and statusChecker. 
In that case, we should set lastHeartbeat to time.now() when starting the status checker initially: https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/pkg/kvstore/etcd.go#L1031
Successful detection of connection problem should not depend on the fact if we receive any event or not. 
Does it make sense?